l1-contracts/contracts/state-transition/StateTransitionManager.sol

```diff
@@ -0,0 +1,240 @@
+
+    /// @dev initialize
+    function initialize(
+        StateTransitionManagerInitializeData calldata _initializeData
+    ) external reentrancyGuardInitializer {
+        require(_initializeData.governor != address(0), "StateTransition: governor zero");
+        _transferOwnership(_initializeData.governor);
+
+        genesisUpgrade = _initializeData.genesisUpgrade;
+        protocolVersion = _initializeData.protocolVersion;
+
+        // We need to initialize the state hash because it is used in the commitment of the next batch
+        IExecutor.StoredBatchInfo memory batchZero = IExecutor.StoredBatchInfo(
+            0,
+            _initializeData.genesisBatchHash,
+            _initializeData.genesisIndexRepeatedStorageChanges,
+            0,
+            EMPTY_STRING_KECCAK,
+            DEFAULT_L2_LOGS_TREE_ROOT_HASH,
+            0,
+            _initializeData.genesisBatchCommitment
+        );
+        storedBatchZero = keccak256(abi.encode(batchZero));
+
+        initialCutHash = keccak256(abi.encode(_initializeData.diamondCut));
+
+        // While this does not provide a protection in the production, it is needed for local testing
+        // Length of the L2Log encoding should not be equal to the length of other L2Logs' tree nodes preimages
+        assert(L2_TO_L1_LOG_SERIALIZE_SIZE != 2 * 32);
+    }
+
+    /// @dev set initial cutHash
+    function setInitialCutHash(Diamond.DiamondCutData calldata _diamondCut) external onlyOwner {
+        initialCutHash = keccak256(abi.encode(_diamondCut));
+    }
+
+    /// @dev set New Version with upgrade from old version
+    function setNewVersionUpgrade(
+        Diamond.DiamondCutData calldata _cutData,
+        uint256 _oldProtocolVersion,
+        uint256 _newProtocolVersion
+    ) external onlyOwner {
+        upgradeCutHash[_oldProtocolVersion] = keccak256(abi.encode(_cutData));
+        protocolVersion = _newProtocolVersion;
+    }
+
+    /// @dev set upgrade for some protocolVersion
+    function setUpgradeDiamondCut(
+        Diamond.DiamondCutData calldata _cutData,
+        uint256 _oldProtocolVersion
+    ) external onlyOwner {
+        upgradeCutHash[_oldProtocolVersion] = keccak256(abi.encode(_cutData));
+    }
+
+    /// upgrade a specific chain
+    function upgradeChainFromVersion(
+        uint256 _chainId,
+        uint256 _oldProtocolVersion,
+        Diamond.DiamondCutData calldata _cutData
+    ) external onlyChainGovernor(_chainId) {
+        bytes32 cutHashInput = keccak256(abi.encode(_cutData));
+        require(cutHashInput == upgradeCutHash[_oldProtocolVersion], "StateTransition: cutHash mismatch");
+
+        IZkSyncStateTransition stateTransitionContract = IZkSyncStateTransition(stateTransition[_chainId]);
+        require(
+            stateTransitionContract.getProtocolVersion() == _oldProtocolVersion,
+            "StateTransition: protocolVersion mismatch in STC when upgrading"
+        );
+        stateTransitionContract.executeUpgrade(_cutData);
```

Comment:
Should we make "soft check" that `stateTransitionContract .getProtocolVersion()` increased?

Reply:
i can add that check here, what do you mean by "soft check"?